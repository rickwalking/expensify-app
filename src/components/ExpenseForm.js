import React from 'react';

import moment from 'moment';

import { SingleDatePicker } from 'react-dates';

export class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: this.props.expense ? this.props.expense.description : '',
            note: this.props.expense ? this.props.expense.note : '',
            amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
            createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: '',
        };
    }

    onDescriptionChange = (event) => {
        const description = event.target.value;
        this.setState(() => ({ description }));
    }

    onNoteChange = (event) => {
        const note = event.target.value;
        this.setState(() => ({ note }));
    }

    onAmountChange = (event) => {
        const amount = event.target.value;

        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }

    onDateChange = (createdAt) => {
        if (!createdAt) {
            return;
        }

        this.setState(() => ({ createdAt }));
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }

    onSubmit = (event) => {
        event.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({
                error: 'Please provide description and amount.',
            }));

            return;
        }

        this.setState(() => ({ error: '' }));

        this.props.onSubmit({
            description: this.state.description,
            note: this.state.note,
            amount: +this.state.amount * 100,
            createdAt: parseFloat(this.state.createdAt.valueOf()),
        });
    }

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type='text'
                        placeholder='Description'
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}/>
                    <input
                        type='number'
                        placeholder='amount'
                        value={this.state.amount}
                        onChange={this.onAmountChange}/>
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                        >
                    </SingleDatePicker>
                    <textarea
                        placeholder='Add a note for your expense (optional)'
                        value={this.state.note}
                        onChange={this.onNoteChange}>
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}