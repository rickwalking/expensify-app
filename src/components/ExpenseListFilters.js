import React from 'react';

import { DateRangePicker }  from 'react-dates';

import { connect } from 'react-redux';

import {
    setTextFilter,
    sortByAmount,
    sortByDate,
    setStartDate,
    setEndDate,
} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }

    onTextChange = (event) => {
        this.props.setTextFilter(event.target.value);
    }

    onSortChange = (event) => {
        const optionValue = event.target.value;

        if (optionValue === 'amount') {
            this.props.sortByAmount();
            return;
        }

        this.props.sortByDate();
    }

    render() {
        return (
            <div>
                <input type='text' value={this.props.filters.text} onChange={this.onTextChange}/>
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    showClearDates={true}
                    isOutsideRange={() => false}>
                </DateRangePicker>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

const mapStateToProps = (state) => {
    return {
        filters: state.filters,
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);