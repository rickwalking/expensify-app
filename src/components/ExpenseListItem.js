import React from 'react';

import { Link } from 'react-router-dom';

import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';

export const ExpenseListItem = (props) => (
    <div>
        <h3>
            <Link to={`/edit/${props.expense.id}`}>{props.expense.description}</Link>
        </h3>
        <p>{props.expense.amount} - {props.expense.createdAt}</p>
        <button onClick={() => {
            props.dispatch(removeExpense({ id: props.expense.id }));
        }}>Remove</button>
    </div>
);

export default connect()(ExpenseListItem);