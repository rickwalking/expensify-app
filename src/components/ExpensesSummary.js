import React from 'react';

import { Link } from 'react-router-dom';

import { getExpensesTotal } from '../selectors/expenses-total';
import { expensesCount } from '../selectors/expenses-count';

import numeral from 'numeral';

import { connect } from 'react-redux';

export const ExpensesSummary = (props) => (
    <div className="page-header">
        <div className="content-container">
            {props.expensesCount > 0 ? (
            <h1 className="page-header__title">
                Viewing <span>{props.expensesCount}</span> {props.expensesCount > 1 ? ' expenses ' : ' expense '} totalling <span>{numeral(props.expensesTotal).format('$0,0.00')}</span>
            </h1>
            ): (
                <h1>You have no expenses yet.</h1>
            )}
            <div className="page-header__actions">
                <Link className="button" to="/create">Add Expense</Link>
            </div>
        </div>
    </div>
)

const mapStateToProps = (state) => {
    return {
        expensesTotal: getExpensesTotal(state.expenses) / 100,
        expensesCount: expensesCount(state.expenses),
    };
}

export default connect(mapStateToProps)(ExpensesSummary);