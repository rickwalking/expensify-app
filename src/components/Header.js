import React from 'react';
import { NavLink } from 'react-router-dom';

export const Header = () => (
    <div>
        <header>Expensify App</header>
        <nav>
            <NavLink to="/" activeClassName="is-active" exact={true}>Home</NavLink>
            <NavLink to="/create" activeClassName="is-active">Create</NavLink>
            <NavLink to="/help" activeClassName="is-active">Help</NavLink>
        </nav>
    </div>
);