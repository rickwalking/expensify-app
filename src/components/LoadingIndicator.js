import React from 'react';

export const LoadingIndicator = () => (
    <div className="loader">
        <img className="loader__image" src="/images/loader.gif" />
    </div>
)