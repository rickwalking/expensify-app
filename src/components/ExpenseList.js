import React from 'react';

import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import { getVisibleExpenses } from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {
            props.expenses.length === 0 ? (
                <h2>No expenses</h2>
            ) : (
                props.expenses.map((expense) => (
                <ExpenseListItem
                    expense={expense}
                    key={expense.id}>
                </ExpenseListItem>
            )))
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters),
        filters: state.filters,
    };
}

export default connect(mapStateToProps)(ExpenseList);