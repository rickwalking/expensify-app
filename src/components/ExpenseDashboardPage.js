import React from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

export const ExpenseDashboardPage = () => (
    <div>
        <ExpenseListFilters></ExpenseListFilters>
        <ExpenseList></ExpenseList>
    </div>
);