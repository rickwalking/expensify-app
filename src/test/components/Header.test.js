import React from 'react';

import { shallow } from 'enzyme';

import { Header } from '../../components/Header';

test('should render Header correcly', () => {
    const wrapper = shallow(<Header />);
    expect(wrapper).toMatchSnapshot();
    // const renderer = new ReactShallowRenderer();

    // renderer.render(<Header></Header>);
    // expect(renderer.getRenderOutput()).toMatchSnapshot();
});