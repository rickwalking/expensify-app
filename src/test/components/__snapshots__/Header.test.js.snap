// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correcly 1`] = `
<div>
  <header>
    Expensify App
  </header>
  <nav>
    <NavLink
      activeClassName="is-active"
      exact={true}
      to="/"
    >
      Home
    </NavLink>
    <NavLink
      activeClassName="is-active"
      to="/create"
    >
      Create
    </NavLink>
    <NavLink
      activeClassName="is-active"
      to="/help"
    >
      Help
    </NavLink>
  </nav>
</div>
`;
